### ISSUE TITLE
Doc section: 'Deploying a VirtualBox VM' causes 'Failed to start Apache HTTPD.'

### ISSUE THREAD

**Issue by josdotso** [NixOS/nixops#391]
Following Section 3.1: Deploying a VirtualBox VM of the NixOps doc, there is an issue with Example 3.1 as evidenced by:

Tail end of running 'nixops deploy -d trivial':
```
webserver> starting the following units: alsa-store.service, get-vbox-nixops-client-key.service, kmod-static-nodes.service, network-local-commands.service, network-setup.service, nix-daemon.service, nscd.service, ntpd.service, systemd-journald.service, systemd-logind.service, systemd-modules-load.service, systemd-sysctl.service, systemd-tmpfiles-clean.timer, systemd-tmpfiles-setup-dev.service, systemd-udev-settle.service, systemd-udev-trigger.service, systemd-udevd.service, systemd-vconsole-setup.service, virtualbox.service
webserver> A dependency job for get-vbox-nixops-client-key.service failed. See 'journalctl -xe' for details.
webserver> A dependency job for virtualbox.service failed. See 'journalctl -xe' for details.
webserver> the following new units were started: audit.service, dev-disk-by\x2dlabel-nixos.device, httpd.service, init.scope, keys.target
webserver> error: unable to activate new configuration
error: activation of 1 of 1 machines failed (namely on ‘webserver’)
```

After sshing into the 'webserver' host:
```
[root@webserver:~]# journalctl -xe
Jan 26 22:31:05 webserver httpd-pre-start[12348]: mkdir: cannot create directory ‘/nix/store/4c8dcjp0c3znivgiij4hj4za2j0bbfmp-valgrind-3.11.0/share/doc’: Read
Jan 26 22:31:05 webserver systemd[1]: httpd.service: Control process exited, code=exited status=1
Jan 26 22:31:05 webserver systemd[1]: Failed to start Apache HTTPD.
Jan 26 22:31:05 webserver systemd[1]: httpd.service: Unit entered failed state.
Jan 26 22:31:05 webserver systemd[1]: httpd.service: Failed with result 'exit-code'.
Jan 26 22:31:10 webserver systemd[1]: httpd.service: Service hold-off time over, scheduling restart.
Jan 26 22:31:10 webserver systemd[1]: nixops-keys.service: Cannot add dependency job, ignoring: Unit nixops-keys.service is masked.
Jan 26 22:31:10 webserver systemd[1]: Stopped Apache HTTPD.
Jan 26 22:31:10 webserver systemd[1]: Starting Apache HTTPD...
Jan 26 22:31:10 webserver httpd-pre-start[12359]: mkdir: cannot create directory ‘/nix/store/4c8dcjp0c3znivgiij4hj4za2j0bbfmp-valgrind-3.11.0/share/doc’: Read
Jan 26 22:31:10 webserver systemd[1]: httpd.service: Control process exited, code=exited status=1
Jan 26 22:31:10 webserver systemd[1]: Failed to start Apache HTTPD.
Jan 26 22:31:10 webserver systemd[1]: httpd.service: Unit entered failed state.
Jan 26 22:31:10 webserver systemd[1]: httpd.service: Failed with result 'exit-code'.
Jan 26 22:31:15 webserver systemd[1]: httpd.service: Service hold-off time over, scheduling restart.
Jan 26 22:31:15 webserver systemd[1]: nixops-keys.service: Cannot add dependency job, ignoring: Unit nixops-keys.service is masked.
Jan 26 22:31:15 webserver systemd[1]: Stopped Apache HTTPD.
Jan 26 22:31:15 webserver systemd[1]: Starting Apache HTTPD...
Jan 26 22:31:15 webserver httpd-pre-start[12370]: mkdir: cannot create directory ‘/nix/store/4c8dcjp0c3znivgiij4hj4za2j0bbfmp-valgrind-3.11.0/share/doc’: Read
Jan 26 22:31:15 webserver systemd[1]: httpd.service: Control process exited, code=exited status=1
Jan 26 22:31:15 webserver systemd[1]: Failed to start Apache HTTPD.
Jan 26 22:31:15 webserver systemd[1]: httpd.service: Unit entered failed state.
Jan 26 22:31:15 webserver systemd[1]: httpd.service: Failed with result 'exit-code'.
Jan 26 22:31:20 webserver systemd[1]: httpd.service: Service hold-off time over, scheduling restart.
Jan 26 22:31:20 webserver systemd[1]: nixops-keys.service: Cannot add dependency job, ignoring: Unit nixops-keys.service is masked.
Jan 26 22:31:20 webserver systemd[1]: Stopped Apache HTTPD.
Jan 26 22:31:20 webserver systemd[1]: Starting Apache HTTPD...
Jan 26 22:31:20 webserver httpd-pre-start[12381]: mkdir: cannot create directory ‘/nix/store/4c8dcjp0c3znivgiij4hj4za2j0bbfmp-valgrind-3.11.0/share/doc’: Read
Jan 26 22:31:20 webserver systemd[1]: httpd.service: Control process exited, code=exited status=1
Jan 26 22:31:20 webserver systemd[1]: Failed to start Apache HTTPD.
Jan 26 22:31:20 webserver systemd[1]: httpd.service: Unit entered failed state.
Jan 26 22:31:20 webserver systemd[1]: httpd.service: Failed with result 'exit-code'.
Jan 26 22:31:26 webserver systemd[1]: httpd.service: Service hold-off time over, scheduling restart.
Jan 26 22:31:26 webserver systemd[1]: nixops-keys.service: Cannot add dependency job, ignoring: Unit nixops-keys.service is masked.
Jan 26 22:31:26 webserver systemd[1]: Stopped Apache HTTPD.
Jan 26 22:31:26 webserver systemd[1]: Starting Apache HTTPD...
Jan 26 22:31:26 webserver httpd-pre-start[12392]: mkdir: cannot create directory ‘/nix/store/4c8dcjp0c3znivgiij4hj4za2j0bbfmp-valgrind-3.11.0/share/doc’: Read
Jan 26 22:31:26 webserver systemd[1]: httpd.service: Control process exited, code=exited status=1
Jan 26 22:31:26 webserver systemd[1]: Failed to start Apache HTTPD.
Jan 26 22:31:26 webserver systemd[1]: httpd.service: Unit entered failed state.
Jan 26 22:31:26 webserver systemd[1]: httpd.service: Failed with result 'exit-code'.
```

Example 3.1 results in httpd-pre-start attempting to create a directory at a read-only location.  Trying this expression with a destination of '/mnt/foobar' works -- but may degrade later steps in the NixOps doc.

Historical reference: https://groups.google.com/forum/#!topic/nixops-users/vDE67k480uM

Thanks!
-Joshua

**Comment by ryanartecona** [comment]
I'm getting this error as well. I've seen it consistently for a couple days now, on every fresh nixops-deployed virtualbox machine. I'm on VirtualBox 5.0.14.

Looks like the virtualbox.service and get-nixops-client-key.service fail because a mutual dependency, dev-vboxguest.device, fails. My `journalctl -xe` after a failed deploy:

```
Jan 27 16:42:01 buggy nixos[1311]: switching to system configuration /nix/store/srrp8lkk7damhpclf1nqlc29c0qawv34-nixos-system-buggy-16.03pre-git
Jan 27 16:42:01 buggy systemd[1]: Stopped target Local File Systems.
Jan 27 16:42:01 buggy systemd[1]: Stopped target All Network Interfaces.
Jan 27 16:42:01 buggy systemd[1]: Stopped target Remote File Systems.
Jan 27 16:42:01 buggy systemd[1]: Reloading.
Jan 27 16:42:01 buggy systemd[1]: Failed to enumerate devices: Invalid argument
Jan 27 16:42:01 buggy systemd[1]: Reached target Timers.
Jan 27 16:42:01 buggy systemd[1]: Reached target System Initialization.
Jan 27 16:42:01 buggy systemd[1]: Reached target Swap.
Jan 27 16:42:01 buggy systemd[1]: Reached target Sockets.
Jan 27 16:42:01 buggy systemd[1]: Reached target Slices.
Jan 27 16:42:01 buggy systemd[1]: Reached target Paths.
Jan 27 16:42:01 buggy systemd[1]: Reached target All Network Interfaces.
Jan 27 16:42:01 buggy systemd[1]: Reached target Remote File Systems.
Jan 27 16:42:01 buggy systemd[1]: Reached target Local File Systems.
Jan 27 16:42:01 buggy systemd[1]: Reached target Services Requiring IP Connectivity.
Jan 27 16:42:01 buggy systemd[1]: Reached target Login Prompts.
Jan 27 16:42:01 buggy systemd[1]: Reached target All Encrypted Links.
Jan 27 16:42:01 buggy systemd[1]: Reached target Basic System.
Jan 27 16:43:31 buggy systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start timed out.
Jan 27 16:43:31 buggy systemd[1]: Timed out waiting for device dev-vboxguest.device.
Jan 27 16:43:31 buggy systemd[1]: Dependency failed for Get NixOps SSH Key.
Jan 27 16:43:31 buggy systemd[1]: get-vbox-nixops-client-key.service: Job get-vbox-nixops-client-key.service/start failed with result 'dependency'.
Jan 27 16:43:31 buggy systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start failed with result 'timeout'.
Jan 27 16:43:31 buggy systemd[1]: Reached target Multi-User System.
Jan 27 16:43:31 buggy nixos[1311]: finished switching to system configuration /nix/store/srrp8lkk7damhpclf1nqlc29c0qawv34-nixos-system-buggy-16.03pre-git
```

And also:
```
[root@buggy:~]# systemctl list-units --type device --all
UNIT                              LOAD   ACTIVE   SUB  DESCRIPTION
dev-disk-by\x2dlabel-nixos.device loaded inactive dead VBOX_HARDDISK nixos
dev-vboxguest.device              loaded inactive dead dev-vboxguest.device

LOAD   = Reflects whether the unit definition was properly loaded.
ACTIVE = The high-level unit activation state, i.e. generalization of SUB.
SUB    = The low-level unit activation state, values depend on unit type.

2 loaded units listed.
```

**Comment by domenkozar** [comment]
@ryanartecona can you paste`journalctl -u dev-vboxguest.device`

**Comment by ryanartecona** [comment]
When I get the error, the journal for that one device isn't very helpful.

```
[root@buggy:~]# journalctl -u dev-vboxguest.device -xef
-- Reboot --
Jan 27 19:13:29 nixos systemd[1]: Found device /dev/vboxguest.
Jan 27 19:18:03 buggy systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start timed out.
Jan 27 19:18:03 buggy systemd[1]: Timed out waiting for device dev-vboxguest.device.
Jan 27 19:18:03 buggy systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start failed with result 'timeout'.
```

**Comment by ryanartecona** [comment]
@domenkozar If it helps, here's [a full journal](https://gist.github.com/ryanartecona/f140006b337c4d9278dc) for a fresh, barebones virtualbox instance created with:

```nix
    { buggy = {
        deployment.targetEnv = "virtualbox";
        deployment.virtualbox.headless = true;
        deployment.virtualbox.memorySize = 1024;
        virtualisation.virtualbox.guest.enable = true;
    };}
```

You can find the errors that show up in red for me by doing a find for `[0;1;31m` on that page. The ones about the failure to rename eth{0,1} to enp0s{3,8} had been happening for me before, and I think they are tangential. I should probably file a separate issue for that. Everything else is Greek to me.

**Comment by ryanartecona** [comment]
Also also, I've noticed that this causes a fresh deploy to hang for ~1min 30s (the timeout for dev-vboxguest.device from `systemctl show dev-vboxguest.device`) late in the deploy steps, and then fail, but after rebooting, a deploy to that same instance will hang the same way and then _succeed_ (nixops shows success), even though that device still shows as dead in systemctl.

Even weirder, a simple reboot usually works and afterward the dev-vboxguest.device shows as healthy, along with a bunch of other devices that were able to start up. I think it's only after a deploy that things go bad. Maybe something in one of the activation scripts is triggering it? I'm kind of at a loss for how to dig deeper.

**Comment by ryanartecona** [comment]
I've been poking at this when I've had time between other things. I haven't made a lot of progress, but I've at least found a workaround.

After digging through a bunch of the activation scripts in nixpkgs/nixos and figuring out how to set `systemd.extraConfig = "LogLevel=debug";` in my configuration, I've been able to pin down a trigger. After a fresh reboot (reboot or stop/start, not a deploy), all the devices in `systemctl list-units -t device -a` show up as healthy. A `systemctl daemon-reload` will reduce the list of healthy devices to only dev-vboxguest.device and dev-disk-by-label-nixos.device, both dead, as I pasted initially above. Also, that daemon-reload will print a `Failed to enumerate devices: Invalid argument` error to the journal before the bunch of "xxx.target changed dead -> active" stuff as it detects things.

The contents of the script in [the get-vbox-nixops-client-key service](https://github.com/NixOS/nixops/blob/master/nix/virtualbox-image-nixops.nix#L21-L29) run just fine even  when the *.device units look dead to systemd, so the workaround is to just add the following to my configuration.

```nix
systemd.services.get-vbox-nixops-client-key.requires = mkForce [];
systemd.services.get-vbox-nixops-client-key.after = mkForce [];
```

**Comment by mboes** [comment]
Seeing this too, only on occasion.

```
$ nixops --version
NixOps 1.3.1pre1673_a0d5681
$ VirtualBox --version
5.0.12_OSE r10481
```

**Comment by djahandarie** [comment]
Also seeing this. @ryanartecona's workaround doesn't seem to have any effect for me. NixOps 1.3 & VirtualBox 4.3.32_Debian.

**Comment by HittingSmoke** [comment]
Also getting this on Ubuntu 15.10 host.
VBox 4.0.14. 
nixops 1.3.1

Running deploy a second time so far always reports success.

**Comment by Gabriel439** [comment]
I'm seeing this error, too.  Running a second deploy fixes the problem

VirtualBox: 5.0.14
nixops: 1.3.1

**Comment by ericsagnes** [comment]
Same problem here with nixops 1.3.1 & virtualbox 5.0.14.

 log:

```
webserver> starting the following units: alsa-store.service, get-vbox-nixops-client-key.service, kmod-static-nodes.service, network-local-commands.service, network-setup.service, nix-daemon.service, nscd.service, ntpd.service, systemd-journald.service, systemd-logind.service, systemd-modules-load.service, systemd-sysctl.service, systemd-tmpfiles-clean.timer, systemd-tmpfiles-setup-dev.service, systemd-udev-settle.service, systemd-udev-trigger.service, systemd-udevd.service, systemd-vconsole-setup.service, virtualbox.service
webserver> A dependency job for get-vbox-nixops-client-key.service failed. See 'journalctl -xe' for details.
webserver> A dependency job for virtualbox.service failed. See 'journalctl -xe' for details.
webserver> the following new units were started: audit.service, dev-disk-by\x2dlabel-nixos.device, init.scope, proc-sys-fs-binfmt_misc.automount
webserver> error: unable to activate new configuration
```

Running the deployment a second time take a quite some of time in `setting up /etc...` but the deployment success.

**Comment by spinus** [comment]
Workaround that worked for me (2 services depend on that device):
```
systemd.services.virtualbox.requires = pkgs.lib.mkForce [];                 
systemd.services.virtualbox.after= pkgs.lib.mkForce [];                     
systemd.services.get-vbox-nixops-client-key.requires = pkgs.lib.mkForce []; 
systemd.services.get-vbox-nixops-client-key.after = pkgs.lib.mkForce []; 
```

Just to notice, that didn't work:
```
systemd.services.dev-vboxguest.serviceConfig.TimeoutSec="1";                
systemd.services.dev-vboxguest.serviceConfig.RuntimeMaxSec="1";             
systemd.services.dev-vboxguest.serviceConfig.Restart="no";                  
```

**Comment by domenkozar** [comment]
@Gabriel439 @spinus what is you base OS?

**Comment by domenkozar** [comment]
Same question for @josdotso and @ericsagnes

**Comment by josdotso** [comment]
@domenkozar  OS X 10.10.5 (14F27)

**Comment by ericsagnes** [comment]
@domenkozar NixOS 16.09pre83955.cc41cb3 (Flounder)

**Comment by domenkozar** [comment]
Interesting, I can't even boot the configuration @ryanartecona posted:

```
buggy> creating VirtualBox VM...
buggy> Virtual machine 'nixops-8f06738f-234c-11e6-982e-54ee7542395c-buggy' is created and registered.
buggy> UUID: 68d37813-d26b-4bac-88c8-f8bd0faf6c1b
buggy> Settings file: '/home/ielectric/VirtualBox VMs/nixops-8f06738f-234c-11e6-982e-54ee7542395c-buggy/nixops-8f06738f-234c-11e6-982e-54ee7542395c-buggy.vbox'
buggy> creating disk ‘disk1’...
buggy> these derivations will be built:
buggy>   /nix/store/r6c9g5w9cswnwy90pxidaij6dx6i4cdm-virtualbox-nixops-15.09.495.4e18cdd.vdi.drv
buggy> building path(s) ‘/nix/store/jnwm4y72jyv5wqlprh10crkwvmv761g0-virtualbox-nixops-15.09.495.4e18cdd.vdi’


buggy> 0%...10%...20%...30%...40%...50%...60%...70%...80%...90%...100%
buggy> Clone medium created in format 'VDI'. UUID: b0af8339-b406-4b38-aa3f-e12f1bd6e1a7
buggy> attaching disk ‘disk1’...
buggy> Waiting for VM "nixops-8f06738f-234c-11e6-982e-54ee7542395c-buggy" to power on...
buggy> VM "nixops-8f06738f-234c-11e6-982e-54ee7542395c-buggy" has been successfully started.
error: command ‘['VBoxManage', 'startvm', u'nixops-8f06738f-234c-11e6-982e-54ee7542395c-buggy', '--type', 'headless']’ failed on machine ‘buggy’ (exit code 1)
```

**Comment by spinus** [comment]
16.03pre75064.faae09f

**Comment by domenkozar** [comment]
Cool, I can reproduce now:

```
buggy> A dependency job for virtualbox.service failed. See 'journalctl -xe' for details.
buggy> A dependency job for get-vbox-nixops-client-key.service failed. See 'journalctl -xe' for details.
buggy> the following new units were started: audit.service, dev-disk-by\x2dlabel-nixos.device, init.scope
buggy> error: unable to activate new configuration
error: activation of 1 of 1 machines failed (namely on ‘buggy’)
```

**Comment by domenkozar** [comment]
@edolstra I'll need your help here.

The issue is that `vboxguest.device` doesn't come up:

```
# systemctl status vboxguest.device                                                                                                                                                                                                            
● vboxguest.device
   Loaded: loaded
   Active: inactive (dead)
```

This is really just a fancy sysytemd way to depend on `/dev/vboxguest` (which does exist).

So maybe local udevrules are not picked up?

```
# grep -R vboxguest /etc/udev/
/etc/udev/rules.d/99-local.rules:# Allow systemd dependencies on vboxguest.
/etc/udev/rules.d/99-local.rules:KERNEL=="vboxguest", TAG+="systemd"
```

**Comment by domenkozar** [comment]
cc @aszlig

**Comment by Gabriel439** [comment]
I was running OS X

**Comment by domenkozar** [comment]
This appears to be broken since systemd upgrade in 16.03. In 15.09 it still worked:

```
# systemctl status dev-vboxguest.device
● dev-vboxguest.device - /dev/vboxguest
   Follow: unit currently follows state of sys-devices-virtual-misc-vboxguest.device
   Loaded: loaded
   Active: active (plugged) since Thu 2016-06-02 10:04:46 UTC; 4min 11s ago
   Device: /sys/devices/virtual/misc/vboxguest

Jun 02 10:04:46 nixos systemd[1]: Found device /dev/vboxguest.
```

**Comment by domenkozar** [comment]
Observations:

- it also fails on `master` branch
- we configure kernel config correctly given [upstream systemd requirements](https://cgit.freedesktop.org/systemd/systemd/tree/README)
- could be we have some outdates package in the chain
- udev is doing it's job correctly
- it's either kernel or systemd

**Comment by domenkozar** [comment]
I've reported this upstream to systemd: https://github.com/systemd/systemd/issues/3423

**Comment by edolstra** [comment]
The vboxguest issue should be fixed, but I'm not sure if it had anything to do with the original report (i.e. the failure to start Apache).

**Comment by domenkozar** [comment]
@josdotso can you reproduce again on 16.03 branch?

**Comment by domenkozar** [comment]
@josdotso using NixOps 1.3.1 and NixOS 16.03 I can't reproduce your problem. If this was 15.09 issue, it's not supported anymore.

Meanwhile `vboxguest` issue was fixed and once NixOps 1.4 is released (any day now), the issue will go away.

Please reopen if anything reappears again or if I missed something.

**Comment by josdotso** [comment]
@domenkozar Sounds good.  I am pretty certain I was trying 15.09 at that time. Thanks for the great support!

**Comment by spinus** [comment]
@domenkozar very nice debugging. Thank you for solving this.

**Comment by domenkozar** [comment]
You should thank @edolstra for solving it - I did spend quite some time debugging it :)

**Comment by spinus** [comment]
Ah, sorry, that wasn't clear for me from a conversation here.
Thank you @edolstra for solving it and thank you @domenkozar for your time on debugging this :-)

**Comment by silverbp** [comment]
Are there directions on how to install nixos 16.03 with nixops on virtualbox? I mean I could create the machine manually I guess or import the downloadable image and let nixops manage it after it's created, but I was hoping it would create it with the 16.03 version.

**Comment by gilligan** [comment]
I guess I am experiencing this right now with https://github.com/NixOS/nixops/issues/473 

> Meanwhile vboxguest issue was fixed and once NixOps 1.4 is released (any day now), the issue will go away.

@domenkozar One month later and NixOps `1.4` is still not out. Unless this is supposed to be fixed with `1.3.1` - which it is not. See #473

**Comment by gilligan** [comment]
@domenkozar I tried with nixops with master and still have to run deploy twice.
